### PR TITLE
[#87250330] sign the user in via token

### PR DESCRIPTION
**Pull request by afeld**
When the approver clicks approve/reject from the notification email, 1019568 now logs that user in to the system via the token. With 3552747, they now get redirected to the normal cart show page, instead of rendering a different view at a different URL.

Checked w/ @cmc333333 and @NoahKunin [in #cybersec](https://18f.slack.com/archives/cybersec/p1422567486000148) to verify that this is ok, security-wise.

### CODE REVIEW

File: spec/controllers/communicarts_controller_spec.rb [hunk 1]
```diff
@@ -265,21 +263,23 @@
...
       end
 
       it 'successfully validates the user_id and cart_id with the token' do
-        approver
-        expect_any_instance_of(Approval).to receive(:update_attributes)
         expect { put 'approval_response', @json_approval_params_with_token }.not_to raise_error
       end
 
+      it "signs the user in" do
+        put 'approval_response', @json_approval_params_with_token
```
Comment (cmc333333) [inline spec/controllers/communicarts_controller_spec.rb]: debugging code

File: spec/controllers/communicarts_controller_spec.rb [hunk 2]
```diff
@@ -303,7 +303,17 @@
         expect { put 'approval_response', @json_approval_params_with_token }.to raise_error(AuthenticationError)
       end
 
-      skip 'marks a token as used'
+      it 'marks a token as used' do
+        expect(ApiToken).to receive(:find_by).with(access_token: "5a4b3c2d1ee1d2c3b4a5").and_return(token)
+        expect_any_instance_of(Approval).to receive(:update_attributes)
+        approver
+
+        put 'approval_response', @json_approval_params_with_token
```
Comment (cmc333333) [inline spec/controllers/communicarts_controller_spec.rb]: debugging code

File: spec/controllers/communicarts_controller_spec.rb (hunk 1 above)
Comment (afeld) [inline spec/controllers/communicarts_controller_spec.rb]: Ha, no, `put` as in "HTTP PUT".

http://api.rubyonrails.org/classes/ActionController/TestCase/Behavior.html#method-i-put

Comment (cmc333333) [inline spec/controllers/communicarts_controller_spec.rb]: My bad!

File: spec/controllers/communicarts_controller_spec.rb [hunk 3]
```diff
@@ -303,7 +303,17 @@
...
+
+        expect(response).to redirect_to(cart_path(cart))
+        token.reload
+        expect(token.used_at).to_not eq(nil)
```
Comment (cmc333333) [inline spec/controllers/communicarts_controller_spec.rb]: Might be worth extending this just a bit further to show that if a user tries to log in with this again, they cannot. I think that's implied by the token, but I'm not sure.

Related question - if a user finds the same email again, and they click one of the links, can they still get to the cart (after logging in)?

**Comment by cmc333333** [comment]
I don't understand all the ruby/rails-isms but the token-based login makes sense to me

**Comment by phirefly** [comment]
Would love to see a browser spec

**Comment by phirefly** [comment]
Great! Just the one comment. Then :shipit:

**Pull request merged**